### ISSUE TITLE
test resources missing after clean - possible Gradle lifecycle issue

### ISSUE THREAD

**Issue by lnhrdt** [spring-cloud/spring-cloud-contract#310]
In my Gradle project, spring cloud contract tests always fail once after clean.

```bash
./gradlew clean
./gradlew latte:test (fails ❌)
./gradlew latte:test (passes ✅)
```

example project: https://github.com/lnhrdt/coffee.git
use commit: 14fe147013

```
.../coffee/latte/build/generated-test-sources/contracts/org/springframework/cloud/contract/verifier/tests/friends/RestTest.java:7: error: package io.leonhardt.coffee.latte.contracts does not exist
import io.leonhardt.coffee.latte.contracts.FriendsRestBase;
                                          ^
.../coffee/latte/build/generated-test-sources/contracts/org/springframework/cloud/contract/verifier/tests/friends/RestTest.java:14: error: cannot find symbol
public class RestTest extends FriendsRestBase {
                              ^
  symbol: class FriendsRestBase
```

Any ideas?

**Comment by marcingrzejszczak** [comment]
That's superstrange... I'll try to check it out this week. You could check out if anything gets generated when you run it for the second time. Also can you show the whole flow of tasks?

**Comment by lnhrdt** [comment]
> Also can you show the whole flow of tasks?

sure thing:

```
./gradlew clean latte:test
:latte:clean
:latte:compileKotlin
Using kotlin incremental compilation
:latte:compileJava NO-SOURCE
:latte:compileGroovy NO-SOURCE
:latte:copyMainKotlinClasses
:latte:processResources
:latte:classes
:latte:compileTestKotlin
Using kotlin incremental compilation
:latte:copyContracts
:latte:generateContractTests
:latte:compileTestJava NO-SOURCE
:latte:compileTestGroovy
.../coffee/latte/build/generated-test-sources/contracts/org/springframework/cloud/contract/verifier/tests/friends/RestTest.java:7: error: package io.leonhardt.coffee.latte.contracts does not exist
import io.leonhardt.coffee.latte.contracts.FriendsRestBase;
                                          ^
.../coffee/latte/build/generated-test-sources/contracts/org/springframework/cloud/contract/verifier/tests/friends/RestTest.java:14: error: cannot find symbol
public class RestTest extends FriendsRestBase {
                              ^
  symbol: class FriendsRestBase
2 errors
startup failed:
Compilation failed; see the compiler error output for details.

1 error

:latte:compileTestGroovy FAILED
:latte:copyTestKotlinClasses

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':latte:compileTestGroovy'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 4.581 secs
```

**Comment by marcingrzejszczak** [comment]
Could you run this twice (the negative and then the positive scenario) with `--debug` flag and attach the output?

**Comment by lnhrdt** [comment]
Sure thing!

```
TERM=dumb ./gradlew clean latte:test --debug &> 1.log
TERM=dumb ./gradlew latte:test --debug &> 2.log
```

[1.log](https://github.com/spring-cloud/spring-cloud-contract/files/1039252/1.log.txt)
[2.log](https://github.com/spring-cloud/spring-cloud-contract/files/1039251/2.log.txt)

**Comment by lnhrdt** [comment]
Aha! On [this line](https://github.com/spring-cloud/spring-cloud-contract/blob/84b0cc6478defca096da30892d8f628357786402/spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/SpringCloudContractVerifierGradlePlugin.groovy#L110) the `generateContractTests` task is configured to depend on `compileTestJava` but I'm writing in Kotlin, which has the task `compileTestKotlin`. Perhaps we should change it to depend on the higher-level `testClasses`.

**Comment by marcingrzejszczak** [comment]
ha! Good point!

**Comment by marcingrzejszczak** [comment]
Care for a PR? :)

**Comment by lnhrdt** [comment]
Already working on it! Trying to figure out how to build the gradle plugin locally and use it in my project to validate the change. If you have any tips let me know. Pausing now, I'm going to dig in again around 1700 CDT.

**Comment by marcingrzejszczak** [comment]
Sure... go to the plugin folder and type `./gradlew clean build install`. That way you'll install the latest snapshots in your .m2

**Comment by lnhrdt** [comment]
After digging in it seems this is more complex than I understood before. I don't have a solution but figured I'd document my context in case I don't come back to this and someone else wants to pick it up or if I need to remind myself what's going on when I can pick this up again.

**TL;DR**
it's a present incompatibility betwen the [kotlin-gradle-plugin](https://github.com/JetBrains/kotlin/tree/v1.1.2/libraries/tools/kotlin-gradle-plugin) and our spring-cloud-contract-gradle-plugin, when the base contract classes are written in Kotlin.

While Java and Groovy have the tasks `compileJava`, `compileGroovy`, `compileTestJava`, and `compileTestGroovy`, Kotlin has the tasks `compileKotlin`, `copyMainKotlinClasses`, `compileTestKotlin`, and `copyTestKotlinClasses`. These extra `copy*` tasks are attached to the lifecycle by finalizing some related Java task ([source code](https://github.com/JetBrains/kotlin/blob/ae2b4f5ff2472c97d73c30d07bf0438e6da4bf94/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinPlugin.kt#L592)). In the scenario with our spring-cloud-contract-gradle-plugin and my Kotlin project, `compileTestGroovy` needs the `copyTestKotlinClasses` to have run, or else the compiled `*RestBase` classes won't exist. As it stands now, `compileTestGroovy` always runs after `compileTestGroovy` so it fails on the first run after a clean.

**Comment by marcingrzejszczak** [comment]
So for sure the workaround for you is to write

`compileTestGroovy.dependsOn 'copyTestKotlinClasses'`

I'll try to programmatically make this work from inside the plugin

**Comment by marcingrzejszczak** [comment]
I gave this a thought and I don't think I should be doing that in the plugin. You can set the tests to be JUnit tests and then you don't have to the workaround. Until then I guess we should wait for more people to express their opinion on this. I'm kind of afraid that changing the order in every project can lead to some issues...

So for now I'm closing this one. We have a workaround to which we can stick at the moment.
